caddytest/caddytest.go

```diff
@@ -0,0 +1,224 @@
+package caddytest
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+)
+
+var internalHostnames = []string{"a.caddy.local", "b.caddy.local", "c.caddy.local"}
+
+// TestContext stores the context of the testing process
+type TestContext struct {
+	t *testing.T
+}
+
+// Complete called to clean up the request free the mutex
+func (tctx *TestContext) Complete() {
```

Comment:
Go 1.14 has new `t.Cleanup()` which I plan to use on this.

Reply:
Go 1.14 is out now, if you wanted to come back to this! I think you can rebase, v2 branch is already targeting 1.14